src/util/numerics.js

```diff
@@ -0,0 +1,70 @@
+// @flow
+
+import * as C from "./combo";
+
+/**
+ * A library for making type-level assertions about numeric properties;
+ * e.g. that a number is finite, or finite and non-negative, or an integer.
+ *
+ * We should avoid having these methods in numeric hot loops by doing "unsafe"
+ * operations in the hot loop, and then validating on exiting the loop.
+ *
+ * Nonetheless, we avoid unnecessary function calls for performance reasons (i.e.
+ * the validator functions avoid calling each other).
+ */
+
+// A number guaranteed to not be NaN, Infinity, or -Infinity
+export opaque type Finite: number = number;
+export function finite(n: number): Finite {
+  if (!isFinite(n)) {
```

Comment:
In case it’s new to you:

```javascript
Number.isFinite("3"); // false (it's a string, of course)
isFinite("3"); // true...
```

Here we have `n: number`, but the function is exported, so it’s not
quite right if the Flow types are unsound—e.g., if an untyped client
grabs us from NPM. (Even if it were guaranteed safe, I’d recommend using
`Number.isFinite` anyway to avoid having to prove this each time I read
the line.)

Likewise below.